README.rst

```diff
@@ -12,77 +12,105 @@ smart_open — utils for streaming large files in Python
 
 ``smart_open`` is well-tested, well-documented, and has a simple, Pythonic API:
 
 .. code-block:: python
 
-  >>> from smart_open import smart_open
-
+  >>> from smart_open import open
+  >>>
   >>> # stream lines from an S3 object
-  >>> for line in smart_open('s3://mybucket/mykey.txt', 'rb'):
-  ...    print(line.decode('utf8'))
+  >>> for line in open('s3://commoncrawl/robots.txt'):
+  ...    print(line, end='')
+  ...    break
+  User-Agent: *
 
+  >>> gz_path = 'smart_open/tests/test_data/crime-and-punishment.txt.gz'
+  >>>
   >>> # stream from/to compressed files, with transparent (de)compression:
-  >>> for line in smart_open('./foo.txt.gz', encoding='utf8'):
-  ...    print(line)
+  >>> for line in open(gz_path, encoding='utf-8'):
+  ...    print(line[:50])
+  ...    break
+  В начале июля, в чрезвычайно жаркое время, под веч
 
   >>> # can use context managers too:
-  >>> with smart_open('/home/radim/foo.txt.bz2', 'wb') as fout:
-  ...    fout.write(u"some content\n".encode('utf8'))
-
-  >>> with smart_open('s3://mybucket/mykey.txt', 'rb') as fin:
+  >>> bz2_path = gz_path.replace('.gz', '.bz2')
+  >>> with open(gz_path) as fin:
+  ...    with open(bz2_path, 'w') as fout:
+  ...        for line in fin:
+  ...           fout.write(line)
+
+  >>> # can use any IOBase operations, like seek
+  >>> with open('s3://commoncrawl/robots.txt', 'rb') as fin:
   ...     for line in fin:
-  ...         print(line.decode('utf8'))
+  ...         print(line.decode('utf-8'), end='')
+  ...         break
   ...     fin.seek(0)  # seek to the beginning
-  ...     b1000 = fin.read(1000)  # read 1000 bytes
-
-  >>> # stream from HDFS
-  >>> for line in smart_open('hdfs://user/hadoop/my_file.txt', encoding='utf8'):
-  ...     print(line)
+  ...     print(fin.read(4))
+  User-Agent: *
+  0
+  b'User'
 
   >>> # stream from HTTP
-  >>> for line in smart_open('http://example.com/index.html'):
-  ...     print(line)
+  >>> for line in open('http://example.com/index.html'):
+  ...     print(line, end='')
+  ...     break
+  <!doctype html>
 
-  >>> # stream from WebHDFS
-  >>> for line in smart_open('webhdfs://host:port/user/hadoop/my_file.txt'):
-  ...     print(line)
+  >>> # stream from HDFS
+  >> for line in open('hdfs://user/hadoop/my_file.txt', encoding='utf8'):
+  ..     print(line)
 
-  >>> # stream content *into* S3 (write mode):
-  >>> with smart_open('s3://mybucket/mykey.txt', 'wb') as fout:
-  ...     for line in [b'first line\n', b'second line\n', b'third line\n']:
-  ...          fout.write(line)
+  >>> # stream from WebHDFS
+  >> for line in open('webhdfs://host:port/user/hadoop/my_file.txt'):
+  ..     print(line)
+
+  >>> # stream content *into* S3 (write mode), using a custom Session for authentication:
+  >>> import boto3
+  >>> tkwa = dict(session=boto3.Session(profile_name='smart_open'))
```

Comment:
Why `tkwa`? What does that stand for?

Reply:
Also, this example seems overly complicated. In the intro, we want the simplest (and most readable) version, with simple `smart_open(`wb`). No `tkwa` (?).

Reply:
tkwa stands for Transport KeyWord Arguments

This code actually writes to our S3 bucket, so it needs some sort of authentication. We can do authentication in other (hidden) ways for this example, e.g. using environment variables. We can move the example with tkwa further down the list, as it demonstrates how to pass custom parameters for more advanced use cases.

Reply:
Yes, move further down please. Let's keep simple things simple (`smart_open`s main allure).

Reply:
> tkwa stands for Transport KeyWord Arguments

How about `transport_kw`?  `tkwa` sounds a little too magical to me.

Reply:
OK, I will use transport_kwargs (for consistency with X_kwargs used elsewhere in the library).

Reply:
How about t_params? It's less magical and shorter than transport_kw.

Reply:
`t_` too magical again :-) I like names to be indicative of their semantic purpose (rather than syntax type, or optimizing for short length).

`transport_kwargs`? `transport_params`?

Reply:
OK, I think transport_params is the way to go.

We may eventually want to add compressor_params later, so params by itself would be too ambiguous.